Build with a nested target platform fails to "bootstrap"

Trying to build the [nested target platform example project](https://github.com/eclipse/tycho/tree/master/tycho-its/projects/target.references) **in a "pristine" environment** fails with an exception when trying to validate the classpath of the `test.bundle` project.

    [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath (default-validate-classpath) on project test.bundle: Execution default-validate-classpath of goal org.eclipse.tycho:tycho-compiler-plugin:2.7.0-SNAPSHOT:validate-classpath failed: bundleLocation not found: null -> [Help 1]

If the project has ever been built _without_ the nested target platforms (e.g. by using `base.target` as the actual TP in the project), then the problem does not appear, even when using the nested TP.

## Environment
I was originally testing this in the context of the 2.7 release (#651) but the same problem happens also in master and in the current release 2.6.0 (although earlier in the build).

## Steps to reproduce
1. Have the Tycho repo downloaded e.g. at path $X
2. Launch the build in a pristine environment, in my case in a new container with a "clean" Maven 3.8 on JDK11:
    ```bash
    $ podman run --rm -it -v "$X/tycho-its/projects/target.references:/mnt" maven:3.8-jdk-11 mvn -f /mnt clean verify
    ```

The build fails with the mentioned exception, while it should succeed instead.

--- a/tycho-bundles/org.eclipse.tycho.p2.resolver.impl/src/main/java/org/eclipse/tycho/p2/target/TargetDefinitionResolver.java
+++ b/tycho-bundles/org.eclipse.tycho.p2.resolver.impl/src/main/java/org/eclipse/tycho/p2/target/TargetDefinitionResolver.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011, 2020 SAP SE and others.
+ * Copyright (c) 2011, 2022 SAP SE and others.
  * This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License 2.0
  * which accompanies this distribution, and is available at
@@ -14,6 +14,7 @@
  *                          - [Bug 568729] Support new "Maven" Target location
  *                          - [Bug 569060] All ids of target file must be different 
  *                          - [Bug 569481] Support for maven target location includeSource="true" attribute
+ *                          - [Issue #401] Support nested targets
  *******************************************************************************/
 package org.eclipse.tycho.p2.target;
 
@@ -133,6 +134,7 @@ public final class TargetDefinitionResolver {
         Map<String, FileTargetDefinitionContent> fileRepositories = new LinkedHashMap<>();
         Map<String, URITargetDefinitionContent> uriRepositories = new LinkedHashMap<>();
         List<MavenTargetDefinitionContent> mavenLocations = new ArrayList<>();
+        List<TargetDefinitionContent> referencedTargetLocations = new ArrayList<>();
         for (Location locationDefinition : definition.getLocations()) {
             if (locationDefinition instanceof InstallableUnitLocation) {
                 InstallableUnitLocation installableUnitLocation = (InstallableUnitLocation) locationDefinition;
@@ -209,6 +211,7 @@ public final class TargetDefinitionResolver {
                 IQueryResult<IInstallableUnit> result = content.query(QueryUtil.ALL_UNITS,
                         new LoggingProgressMonitor(logger));
                 unitResultSet.addAll(result);
+                referencedTargetLocations.add(content);
             } else {
                 logger.warn("Target location type '" + locationDefinition.getTypeDescription() + "' is not supported");
             }
@@ -236,6 +239,11 @@ public final class TargetDefinitionResolver {
         for (MavenTargetDefinitionContent mavenContent : mavenLocations) {
             metadataRepositories.add(mavenContent.getMetadataRepository());
             artifactRepositories.add(mavenContent.getArtifactRepository());
+        }
+        //preliminary step: add all referenced targets:
+        for (TargetDefinitionContent referenceContent : referencedTargetLocations) {
+            metadataRepositories.add(referenceContent.getMetadataRepository());
+            artifactRepositories.add(referenceContent.getArtifactRepository());
         }
         //now we can resolve the p2 sources
         if (installableUnitResolver != null) {

